### PR TITLE
[RFC] API Routes in Expo Router

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Servers are an important part of developing many different types of apps, but they're much harder to configure than they need to be.

API Routes will enable users to express some abstract JavaScript code that runs in a server by simply creating a file in the app directory, and adding the `+api.js` suffix. For example, to securely interact with OpenAI, simply:

```ts 
// app/generate+api.ts
import { ExpoRequest, ExpoResponse } from 'expo-router/server';

export async function POST(req: ExpoRequest): Promise<ExpoResponse> {
  const { prompt } = await req.json();

  const json = await fetch('https://api.openai.com/v1/engines/text-davinci-003/completions', {
    headers: {
      'Content-Type': 'application/json',
      // `OPENAI_API_KEY` is pulled from the .env file when running in Expo CLI.
      Authorization: `Bearer ${process.env.OPENAI_API_KEY ?? ''}`,
    },
    method: 'POST',
    body: JSON.stringify({
      prompt,
      max_tokens: 100,
  }),
  }).then(res => res.json());

  // Return as JSON
  return ExpoResponse.json(json);
}
```

This will be served at `http://localhost:8081/generate` with `npx expo` and can be used by making a request:

```sh
$ curl -X POST -H "Content-Type: application/json" -d \'{"prompt":"Hello, my name is"}\' http://localhost:8081/generate
```

Expo Router polyfills the URL and `window.location` object on native to allow for universally requesting with a relative URL:

```js
// Expo prepends the host and port to the URL automatically in development.
const json = await fetch('/generate').then(res => res.json());
```

# How

- API Routes are bundled with Metro, leveraging all the same functionality as the rest of the app and website. 
  - The project babel config is used to transpile the API routes. Indication is passed to the Babel caller via the `isServer` boolean. This can be used to change the preset based on the environment.
  - Each API route is bundled into a standalone file in the `dist/_expo` directory.  This is akin to ncc, the tool we use to make Create Expo App download in ~1 second.
- Create a new package `@expo/server` which includes the requisite middleware and runtime polyfills for the Expo server environment.
- Add a new routes manifest which will be used by `@expo/server` to serve up the three types of routes: HTML routes, API routes, and not found routes (404s).
- Add a new export `expo-router/server` (potentially will be moved to `expo/server`) which contains the `ExpoRequest` and `ExpoResponse` objects. These are all based on the WinterCG specification, and include some additional properties for interop with the Expo Router filesystem convention. These are inspired by Remix, SvelteKit, and Next.js for simplicity.
- Add a new export mode `web.output: "server"` which can be used to export a dynamic server. Note: I may drop this for now and make server the default since there's no expo-specific hosting code that must be exported.
- This PR adds the ability to host the app with an express server, different production adapters to follow.

# Test Plan

In addition to all the E2E Metro tests, I've added a new E2E runner which starts a server and pings different requests to ensure expected behavior. These run in the CLI as opposed to the `@expo/server` package.

- resolve ENG-10057 ENG-8243 ENG-8082 ENG-8079 ENG-8242 ENG-8081 ENG-8080 ENG-9625

### CODE REVIEW

File: apps/router-e2e/__e2e__/server/app/api/externals+api.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { ExpoResponse } from 'expo-router/server';
+import fs from 'node:fs';
+import path from 'path';
+
+export function GET(): ExpoResponse {
+  console.log(fs);
+  // const txt = fs.readFileSync(path.join(__dirname, '../../../file.txt'), 'utf8');
+  return new ExpoResponse(path.join('a', 'b', 'c'), { status: 200 });
```
Comment (byCedric) [inline apps/router-e2e/__e2e__/server/app/api/externals+api.ts:8]: This probably should have the `fs.readFileSync` and file, right? Maybe even just try to read the app name, and return it as app name?

File: apps/router-e2e/__e2e__/server/app/methods+api.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { ExpoResponse } from 'expo-router/server';
+
+/** @type {import('expo-router/server').RequestHandler} */
+export function GET() {
```
Comment (byCedric) [inline apps/router-e2e/__e2e__/server/app/methods+api.ts:4]: Maybe also documenting/testing this one could help users with TS.

```suggestion
import { ExpoResponse, type RequestHandler } from 'expo-router/server';

export const GET: RequestHandler = () => {
```

Comment (EvanBacon) [reply apps/router-e2e/__e2e__/server/app/methods+api.ts:4]: This should be covered in the `babel-preset-expo` tests.

File: docs/pages/router/reference/api-routes.mdx [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+Before you deploy with third-parties, it's good to be familiar with the basics of the [`npx expo export`](/more/expo-cli#exporting) command. Specifically:
+
+- `/dist` is the default export directory for Expo CLI.
+- Files in `/public` will be copied to `/dist` on export.
+- The `@expo/server` package is included with `expo` and delegates requests to the server routes.
+- `@expo/server` does **not** inflate environment variables from `.env` files, these are expected to be loaded either by the hosting provider or the user.
+- Metro is not included in the server.
+
+### Express
```
Comment (byCedric) [inline docs/pages/router/reference/api-routes.mdx:208]: Did you test with Bun yet? If not, I'd be happy to write the Bun equivalent as we have for Express.

File: packages/@expo/server/src/environment.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  RequestInfo,
+  RequestInit,
+  Response,
+  ResponseInit,
+  Headers,
+} from '@remix-run/node';
+import { URL } from 'node:url';
+
+import { ExpoRouterServerManifestV1FunctionRoute } from './types';
+
```
Comment (byCedric) [inline packages/@expo/server/src/environment.ts:13]:
```suggestion

declare global {
  const Request: ExpoRequest;
  const ExpoRequest: ExpoRequest;
  const Response: ExpoResponse;
  const ExpoResponse: ExpoResponse;
}

```

File: packages/@expo/server/src/environment.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  installRemixGlobals();
+
+  // @ts-expect-error
+  global.Request = ExpoRequest;
+  // @ts-expect-error
+  global.Response = ExpoResponse;
+  // @ts-expect-error
+  global.ExpoResponse = ExpoResponse;
+  // @ts-expect-error
+  global.ExpoRequest = ExpoRequest;
```
Comment (byCedric) [inline packages/@expo/server/src/environment.ts:25]:
```suggestion
  global.Request = ExpoRequest;
  global.Response = ExpoResponse;
  global.ExpoResponse = ExpoResponse;
  global.ExpoRequest = ExpoRequest;
```

File: packages/@expo/server/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        return null;
+      }
+      return fs.readFileSync(filePath, 'utf-8');
+    },
+    getApiRoute = async (route) => {
+      const filePath = path.join(distFolder, '_expo/functions', route.page + '.js');
+
+      debug(`Handling API route: ${route.page}: ${filePath}`);
+
+      // TODO: What's the standard behavior for malformed projects?
```
Comment (byCedric) [inline packages/@expo/server/src/index.ts:66]: 🤷 Probably just throw an error in CLI, and return 500 as server error with a nice UI?

image

File: packages/expo-router/src/__tests__/getServerManifest.test.node.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  });
+});
+
+xit(`converts single basic`, () => {
+  expect(getServerManifest(getRoutesFor(['./home.js']))).toEqual([
+    { namedRegex: '^/home(?:/)?$', routeKeys: {} },
+  ]);
+});
+
+it(`sorts`, () => {
```
Comment (byCedric) [inline packages/expo-router/src/__tests__/getServerManifest.test.node.ts:42]:
```suggestion
it(`sorts deepest to shallowest paths`, () => {
```

File: docs/pages/router/reference/api-routes.mdx (hunk 3 above)
Comment (EvanBacon) [reply docs/pages/router/reference/api-routes.mdx:208]: We should wait until this lands before adding any additional adapters.

File: packages/@expo/server/src/index.ts (hunk 6 above)
Comment (EvanBacon) [reply packages/@expo/server/src/index.ts:66]: looks good than, but we probably want something more unified in the future.

**Pull request merged**